### ISSUE TITLE
WaitForExit

### ISSUE THREAD

**Issue by fabekoma** [jantari/LSUClient#32]
**Describe the bug**
The process stop during installing a driver (Probably one of this 4) :
g7audc8us17 - Realtek Audio Codec Driver - 10 [64] - Audio
u2etn29us14 - Intel Lan driver - 10 [64] - Networking LAN Ethernet
u8amt62us14 - Intel AMT Driver - 10 [64] - Motherboard Devices core chipset onboard video PCIe switches
u1rau22us17 - Intel RST driver - 10 [64] - Storage

With this error :
No process is associated to this object ( Invoke-PackageCommand.ps1:82:5 )

**Error messages or Screenshots**
![MicrosoftTeams-image (3)](https://user-images.githubusercontent.com/8105679/120482480-7ab14e00-c3b1-11eb-8cf8-c6166fd98c2b.png)

**Please complete the following information:**
 - Computer model: Lenovo M900 (10FL)
 - Module version used: 1.2.4

**Comment by jantari** [comment]
Hey, thanks for reporting an issue.

A few questions about this:

1. Does this error happen consistently / every time?
2. Do you also have the powershell console output from when this happens? It *should* have printed a warning or two prior to the error
3. Can you give me the exact `Install-LSUpdate` command that you run or that is being run by `Deploy-Application.ps1`?

At first it looks like this could only be caused by the self-extracting EXE archive that Lenovo packages its drivers in being corrupted, e.g. it could have been a bad download. It's trying to extract the package, but gets a `ERROR_BAD_EXE_FORMAT` error, then tries a second time with a different method but gets `ERROR_BAD_EXE_FORMAT` again which is why the error is thrown. This would mean the exe file for one of the packages is invalid or corrupted.

With that being said, ideally the script shouldn't halt and exit entirely because of one unreadable or invalid exe. That was an oversight on my part and should be fixed by e47eed78d42861dd896f5447bbba8f39a33f9216.

**Comment by fabekoma** [comment]
This is what i get this morning with the same device :

PS C:\windows\system32> Import-Module -Name C:\temp\LSUClient

$Allupdates = Get-LSUpdate | Where-Object { $_.Installer.Unattended -and $_.category -ne 'Software and Utilities' -and $_.severity -ne 'Optional' }
$Allupdates | select *
$Allupdates | Install-LSUpdate




ID           : g7audc8us17
Title        : Realtek Audio Codec Driver - 10 [64]
Category     : Audio
Version      : 6.0.8924.1
Severity     : Recommended
ReleaseDate  : 07/05/2020 00:00:00
RebootType   : 3
Vendor       : Lenovo
Size         : 204681154
URL          : https://download.lenovo.com/consumer/desktop/g7audc8us17_2_.xml
Extracter    : PackageExtractInfo
Installer    : PackageInstallInfo
IsApplicable : True
IsInstalled  : False

ID           : u2etn29us14
Title        : Intel Lan driver - 10 [64]
Category     : Networking LAN Ethernet
Version      : 12.19.0.16
Severity     : Recommended
ReleaseDate  : 09/02/2021 00:00:00
RebootType   : 0
Vendor       : Lenovo
Size         : 2080797
URL          : https://download.lenovo.com/pccbbs/thinkcentre_drivers/u2etn29us14_2_.xml
Extracter    : PackageExtractInfo
Installer    : PackageInstallInfo
IsApplicable : True
IsInstalled  : False

ID           : u8amt62us14
Title        : Intel AMT Driver - 10 [64]
Category     : Motherboard Devices core chipset onboard video PCIe switches
Version      : 2040.100.0.1029
Severity     : Recommended
ReleaseDate  : 09/02/2021 00:00:00
RebootType   : 3
Vendor       : Lenovo
Size         : 47563519
URL          : https://download.lenovo.com/pccbbs/thinkcentre_drivers/u8amt62us14_2_.xml
Extracter    : PackageExtractInfo
Installer    : PackageInstallInfo
IsApplicable : True
IsInstalled  : False

ID           : fwjyb5usa
Title        : Flash BIOS Updates for ThinkCentre M700 Tiny, M800, M900, M900x Tiny - [32, 64]
Category     : BIOS UEFI
Version      : 0.0.0.0
Severity     : Recommended
ReleaseDate  : 04/02/2021 00:00:00
RebootType   : 1
Vendor       : Lenovo
Size         : 7004788
URL          : https://download.lenovo.com/pccbbs/thinkcentre_bios/fwjyb5usa_2_.xml
Extracter    : PackageExtractInfo
Installer    : PackageInstallInfo
IsApplicable : True
IsInstalled  : False

Save-LSUpdate : Not all packages could be downloaded, the following errors were encountered:
https://download.lenovo.com/pccbbs/thinkcentre_drivers/u8amt62us14_2_.xml : Le nom distant n'a 
pas pu être résolu: 'download.lenovo.com'
https://download.lenovo.com/pccbbs/thinkcentre_drivers/u8amt62us14.exe : Le nom distant n'a pas 
pu être résolu: 'download.lenovo.com'
https://download.lenovo.com/pccbbs/thinkcentre_drivers/u8amt62us14.txt : Le nom distant n'a pas 
pu être résolu: 'download.lenovo.com'
https://download.lenovo.com/pccbbs/thinkcentre_drivers/getw10ver7.exe : Le nom distant n'a pas pu 
être résolu: 'download.lenovo.com'
Au caractère C:\temp\LSUClient\public\Install-LSUpdate.ps1:39 : 17
+                 Save-LSUpdate -Package $PackageToProcess -Path $Path
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Save-LSUpdate
 
WARNING: The file to be run is not an executable - falling back to ShellExecute
WARNING: Running with ShellExecute - process output cannot be captured!
WARNING: The file to be run is not an executable - falling back to ShellExecute
Invoke-PackageCommand : Exception lors de l'appel de «WaitForExit» avec «0» argument(s): «Aucun 
processus n'est associé à cet objet.»
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:66 : 25
+ ...     return (Invoke-PackageCommand -Path:$Path -Command:$Command -Fall ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Invoke-PackageCommand], MethodInvocationExceptio 
   n
    + FullyQualifiedErrorId : InvalidOperationException,Invoke-PackageCommand
 
Impossible d’appeler une méthode dans une expression Null.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:78 : 9
+         $StdOutAsync = $process.StandardOutput.ReadToEndAsync()
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
Impossible d’appeler une méthode dans une expression Null.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:79 : 9
+         $StdErrAsync = $process.StandardError.ReadToEndAsync()
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
Exception lors de l'appel de «WaitForExit» avec «0» argument(s): «Aucun processus n'est associé à 
cet objet.»
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:82 : 5
+     $process.WaitForExit()
+     ~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : InvalidOperationException
 
Impossible d’extraire la variable « $StdOutAsync », car elle n’a pas été définie.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:85 : 30
+         $StdOutInOneString = $StdOutAsync.GetAwaiter().GetResult()
+                              ~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (StdOutAsync:String) [], RuntimeException
    + FullyQualifiedErrorId : VariableIsUndefined
 
Impossible d’extraire la variable « $StdErrAsync », car elle n’a pas été définie.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:86 : 30
+         $StdErrInOneString = $StdErrAsync.GetAwaiter().GetResult()
+                              ~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (StdErrAsync:String) [], RuntimeException
    + FullyQualifiedErrorId : VariableIsUndefined
 
Impossible d’appeler une méthode dans une expression Null.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:88 : 9
+         [string[]]$StdOutLines = $StdOutInOneString.Split(
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
Impossible d’appeler une méthode dans une expression Null.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:93 : 9
+         [string[]]$StdErrLines = $StdErrInOneString.Split(
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
WARNING: The command or file 'C:\Users\Admin\AppData\Local\Temp\LSUPackages\u8amt62us14\\DriverSet
up.exe /VERYSILENT' could not be found from 'C:\Users\Admin\AppData\Local\Temp\LSUPackages\u8amt62
us14' and was not run
WARNING: Installation of package 'u8amt62us14 - Intel AMT Driver - 10 [64]' FAILED - the installati
on could not start
Save-LSUpdate : Not all packages could be downloaded, the following errors were encountered:
https://download.lenovo.com/pccbbs/thinkcentre_bios/fwjyb5usa_2_.xml : Le nom distant n'a pas pu 
être résolu: 'download.lenovo.com'
https://download.lenovo.com/pccbbs/thinkcentre_bios/fwjyb5usa.exe : Le nom distant n'a pas pu 
être résolu: 'download.lenovo.com'
https://download.lenovo.com/pccbbs/thinkcentre_bios/fwjyb5usa.txt : Le nom distant n'a pas pu 
être résolu: 'download.lenovo.com'
Au caractère C:\temp\LSUClient\public\Install-LSUpdate.ps1:39 : 17
+                 Save-LSUpdate -Package $PackageToProcess -Path $Path
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Save-LSUpdate
 
WARNING: The file to be run is not an executable - falling back to ShellExecute
WARNING: Running with ShellExecute - process output cannot be captured!
WARNING: The file to be run is not an executable - falling back to ShellExecute
Invoke-PackageCommand : Exception lors de l'appel de «WaitForExit» avec «0» argument(s): «Aucun 
processus n'est associé à cet objet.»
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:66 : 25
+ ...     return (Invoke-PackageCommand -Path:$Path -Command:$Command -Fall ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Invoke-PackageCommand], MethodInvocationExceptio 
   n
    + FullyQualifiedErrorId : InvalidOperationException,Invoke-PackageCommand
 
Impossible d’appeler une méthode dans une expression Null.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:78 : 9
+         $StdOutAsync = $process.StandardOutput.ReadToEndAsync()
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
Impossible d’appeler une méthode dans une expression Null.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:79 : 9
+         $StdErrAsync = $process.StandardError.ReadToEndAsync()
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
Exception lors de l'appel de «WaitForExit» avec «0» argument(s): «Aucun processus n'est associé à 
cet objet.»
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:82 : 5
+     $process.WaitForExit()
+     ~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : InvalidOperationException
 
Impossible d’extraire la variable « $StdOutAsync », car elle n’a pas été définie.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:85 : 30
+         $StdOutInOneString = $StdOutAsync.GetAwaiter().GetResult()
+                              ~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (StdOutAsync:String) [], RuntimeException
    + FullyQualifiedErrorId : VariableIsUndefined
 
Impossible d’extraire la variable « $StdErrAsync », car elle n’a pas été définie.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:86 : 30
+         $StdErrInOneString = $StdErrAsync.GetAwaiter().GetResult()
+                              ~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (StdErrAsync:String) [], RuntimeException
    + FullyQualifiedErrorId : VariableIsUndefined
 
Impossible d’appeler une méthode dans une expression Null.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:88 : 9
+         [string[]]$StdOutLines = $StdOutInOneString.Split(
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
Impossible d’appeler une méthode dans une expression Null.
Au caractère C:\temp\LSUClient\private\Invoke-PackageCommand.ps1:93 : 9
+         [string[]]$StdErrLines = $StdErrInOneString.Split(
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation : (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
WARNING: Either this is not a BIOS Update or it's an unsupported installer for one, skipping instal
lation!


Add save LSUpdate :

PS C:\windows\system32> $Allupdates | Save-LSUpdate

PS C:\windows\system32> Import-Module -Name C:\temp\LSUClient

$Allupdates = Get-LSUpdate | Where-Object { $_.Installer.Unattended -and $_.category -ne 'Software and Utilities' -and $_.severity -ne 'Optional' }
$Allupdates | select *
$Allupdates | Install-LSUpdate


ID           : u8amt62us14
Title        : Intel AMT Driver - 10 [64]
Category     : Motherboard Devices core chipset onboard video PCIe switches
Version      : 2040.100.0.1029
Severity     : Recommended
ReleaseDate  : 09/02/2021 00:00:00
RebootType   : 3
Vendor       : Lenovo
Size         : 47563519
URL          : https://download.lenovo.com/pccbbs/thinkcentre_drivers/u8amt62us14_2_.xml
Extracter    : PackageExtractInfo
Installer    : PackageInstallInfo
IsApplicable : True
IsInstalled  : False

ID           : fwjyb5usa
Title        : Flash BIOS Updates for ThinkCentre M700 Tiny, M800, M900, M900x Tiny - [32, 64]
Category     : BIOS UEFI
Version      : 0.0.0.0
Severity     : Recommended
ReleaseDate  : 04/02/2021 00:00:00
RebootType   : 1
Vendor       : Lenovo
Size         : 7004788
URL          : https://download.lenovo.com/pccbbs/thinkcentre_bios/fwjyb5usa_2_.xml
Extracter    : PackageExtractInfo
Installer    : PackageInstallInfo
IsApplicable : True
IsInstalled  : False

WARNING: Could not test this ThinkCentre BIOS-Update for the /sccm (suppress reboot) parameter: A problem occured when calling the
 native API 'WriteConsoleInput'. Try running this script in a terminal that supports it, such as the default conhost or anything t
hat builds atop of ConPTY.
Exception lors de l'appel de «Kill» avec «0» argument(s): «Impossible de traiter la demande, car le processus (7296) n'est plus 
exécuté.»
Au caractère C:\temp\LSUClient\private\Test-Wflash2ForSCCMParameter.ps1:134 : 17
+                 $process.Kill()
+                 ~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : InvalidOperationException
 
BIOS UPDATE SUCCESS: An immediate full SHUTDOWN is strongly recommended to allow the BIOS update to complete!


verify updates :

$Allupdates = Get-LSUpdate | Where-Object { $_.Installer.Unattended -and $_.category -ne 'Software and Utilities' -and $_.severity -ne 'Optional' }
$Allupdates

ID        Title                                                                           Category  Version Severity    IsApplica
                                                                                                                        ble      
--        -----                                                                           --------  ------- --------    ---------
fwjyb5usa Flash BIOS Updates for ThinkCentre M700 Tiny, M800, M900, M900x Tiny - [32, 64] BIOS UEFI 0.0.0.0 Recommended True

**Comment by jantari** [comment]
Your root problem is just this:

> Le nom distant n'a pas pu être résolu: 'download.lenovo.com'

I can't say why that's happening for you, but it's a DNS problem on your network - either DNS is timing out because it's too slow or `download.lenovo.com` is (sometimes) being blocked, so some packages cannot be downloaded by LSUClient. `Save-LSUpdate` gives you the links to the files it couldn't download, try and see if you can download them manually and whether that works. Maybe there's a firewall or proxy blocking certain files for you, in that case you'll have to check the logs on those.

**Comment by jantari** [comment]
Hey,

I fixed the "error-sprawl" you were getting and those fixes are now released with Version 1.2.5. This should now be handled more gracefully with only one error or warning.

The underlying problem I'm pretty sure must be somewhere on your network as you don't seem to have uninterrupted/unrestricted access to `download.lenovo.com` and thus some packages couldn't be downloaded properly. I don't think there's anything more I can do here.
